### ISSUE TITLE
How to get predication from model in custom loss function

### ISSUE THREAD

**Issue by JasonRuan5** [Microsoft/CNTK#2292]
I would like to create a custom loss function like below.
def custom_loss_function(z, targets):
    #get pred from model z
    ce = calc_loss_with_weight(pred, targets)
    return ce

Is it possible to get prediction outputs from model z?

**Comment by ebarsoumMS** [comment]
Is you custom loss function a CNTK expression or pure Python code?

**Comment by JasonRuan5** [comment]
Pure Python code is preferred.  CNTK expressions are fine too if custom weights can be used in loss function.  Thanks!

**Comment by ebarsoumMS** [comment]
It is better to use CNTK expression, because it will run on both GPU and CPU, and you will get gradient for free. In this case, nothing special is needed: Here a custom cross entropy (assuming z is the model output before softmax and this is a classification problem.):  

pred = C.softmax(z)
ce = C.negate(C.reduce_sum(C.element_times(targets, C.log(pred)), axis=-1))

Pretty much most numpy OPs are available in CNTK.

**Comment by JasonRuan5** [comment]
Great and thanks!  I hope to create one dimensional array, predN, after pred = C.softmax(z).  All elements of predN are  1 - pred[i], except at nTarget index, which is pred[nTarget], so I did as below:
`def calcNewPred(nTarget, pred) :
    nClass = np.size(pred, axis=0) 
    predN = np.copy(pred)
    for i in range(nClass):
        if (i != nTarget) :
            predN[i] = 1 - pred[i]
    return pfPred`

I got error at predN[i] = 1 - pred[i].  I am new to np and CNTK expression.  I would be greatly appreciated if you could advise me how I could correct that?

**Comment by JasonRuan5** [comment]
If the above code could is not doable, could I have an alternative solution below?
def WeightedError(z, targets):
    nTarget = np.argmax(targets) # return index of the max value of target
    pred = C.softmax(z)
    nPred = np.argmax(pred) # return index of the max value of pred
    if (nTarget == nPred) :
        return 0    #matched
    elif (nTarget == 0) :
        return nPred    #false positive
    elif (nPred == 0) :
        return 2* nTarget   #false negative
    elif (nPred > nTarget):
        return 0.4 * (nPred- nTarget)  #over estimated

    return 0.55 * (nTarget - nPred) #under estimated

I got runtime error when I do that?  What did I do wrong?

**Comment by JasonRuan5** [comment]
Also, is there any way to convert the above pred and targets to array of floats?  I assume that pred and targets are from each train sample, am I right?

**Comment by JasonRuan5** [comment]
I am still struggling with custom weighted loss function.  I've realized that loss function does not just return an error value as it will raise exception if I use CalcWeight below as loss function 
# calculate weight based on predicating and target results 
def CalcWeight(pred, targets):
    nTarget = np.argmax(targets) # return index of the max value of target
    nPred = np.argmax(pred) # return index of the max value of pred
    diff = 0.01
    if (nTarget == nPred) :
        diff = 0.01         #match
    elif (nTarget == 0) :
        diff = 1.0* nPred    #false positive
    elif (nPred == 0) :
        diff = 2.0* nTarget   #false negative
    elif (nPred > nTarget):
        diff = 0.45 * (nPred- nTarget)  #over estimated
    else :
        diff = 0.55 * (nTarget - nPred) #under estimated

    return diff;

def WeightedError(pred, targets):
    weight = CalcWeight(pred, targets) # get weight based on predicating and target results 
    return weight * C.negate(C.reduce_sum(C.element_times(targets, C.log(pred)), axis=-1))

It is fine to use WeightedError(pred, targets) as loss function, but the weight has little effect on train.  My questions are:
1. What is wrong with above functions?
2. What should loss function return?
3. How could we apply the above weights to the loss function?
4. Can we simply convert or cast float weight value to whatever loss function should return, e.g. function pointer?
5. Any custom weight loss example code?

Any advice will be greatly appreciated.

**Comment by ebarsoumMS** [comment]
The problem of the above is that pred and targets are symbol in the graph, you don't have their value. If you want to use pure Python, then you need to implement UserDefine function: https://cntk.ai/pythondocs/cntk.ops.functions.html?highlight=userfunction#cntk.ops.functions.UserFunction

And here couple of examples: https://github.com/Microsoft/CNTK/blob/16d01d7a3256c8bb689ab4ddc1b66ae842c186a9/bindings/python/cntk/ops/tests/userfunction_test.py

Otherwise, you need to implement the above using CNTK expression in this case, you will need to use C.element_select, C.greater...etc (no python if).

**Comment by JasonRuan5** [comment]
Excellent, I think I got it.  Thanks so much for your help, Emad!
